models/cam-fv/camdart_obs_preprocessor.f90

```diff
@@ -0,0 +1,320 @@
+
+
+      ! ----------- MODIFY HERE ----------------
+      ! modify obs_out to match what you need
+
+      ! If you need something in the obs_def type, here's how you do that.
+      ! This just queries the location and then makes a call to unpackage
+      ! the location into an array.
+
+      ! call get_obs_def(obs_out, this_obs_def)
+      ! obs_loc       = get_obs_def_location(this_obs_def)
+      ! obs_loc_array = get_location(obs_loc)
+ 
+      ! If you change something in the observation definition 
+      ! you have to update the actual observation.
+
+      ! call set_obs_def(obs_out, this_obs_def)
+
+      ! if you do NOT want to insert this observation in the output
+      ! file, set a condition that skips the next block.  you *must*
+      ! still call get_next_obs() before looping.
+
+      ! ----------- MODIFY HERE ----------------
+
+      call get_obs_def(obs_out, this_obs_def)
+      obs_loc       = get_obs_def_location(this_obs_def)
+      obs_loc_array = get_location(obs_loc)
+
+      vert_value = obs_loc_array(3)
+      which_vert = int(query_location(obs_loc, "which_vert"))
+
+      call obs_too_high(vert_value, which_vert, too_high)
+      if (verbose) print *, 'vert_value = ', vert_value, ' which_vert = ', which_vert, ' too_high = ', too_high
+
+      if (too_high == 0) then
+
+         ! Since the stride through the observation sequence file is always
+         ! guaranteed to be in temporally-ascending order, we can use the
+         ! 'previous' observation as the starting point to search for the
+         ! correct insertion point.  This speeds up the insert code a lot.
+   
+         if (num_inserted > 0) then
+            call insert_obs_in_seq(seq_out, obs_out, prev_obs_out)
+         else
+            call insert_obs_in_seq(seq_out, obs_out)
+         endif
+   
+         prev_obs_out = obs_out  ! update position in seq for next insert
+         num_inserted = num_inserted + 1
+   
+         if (print_every > 0) then
+            if (mod(num_inserted,print_every) == 0) then
+               print*, 'inserted number ',num_inserted,' of ',size_seq_out
+            endif
+         endif
+
+      endif
+
+      ! ---- THIS MUST BE EXECUTED BEFORE THE NEXT LOOP ----
+
+      call get_next_obs(seq_in, obs_in, next_obs_in, is_this_last)
+
+   enddo ObsLoop
+
+else
+   write(msgstring, *)'no first observation in ',trim(filename_in)
+   call error_handler(E_MSG,'main', msgstring,source)
+endif
+
+if (.not. verbose) then
+   print*, '---------  Obs seqs '
+   print*, 'Number of obs input sequence    :         ', size_seq_in
+   print*, 'Number of obs copied to output  :         ', num_inserted
+   print*, '---------------------------------------------------------'
+endif
+
+write(msgstring, *) 'Starting to process output sequence file ', &
+                            trim(filename_out)
+call error_handler(E_MSG,'main',msgstring,source)
+
+print*, 'Number of obs in the output seq file :', get_num_key_range(seq_out)
```

Comment:
Why the different methods for printing messages? (print and error_handler(E_MSG...)

Reply:
line 271 looks like a debugging line that should be removed.    i'm also a bit mystified about the  "if not verbose" test at 260.  it looks like that should be "if verbose".  i think the idea was verbose messages might print to the screen but other messages go into the log as well as screen.  my suggestion is to run it a couple times with and without verbose and decide what you think are the most useful messages for various situations - like if the output file is empty, what info would help to print out?  and put those prints in the verbose sections.  and if you want to use msg for everything, i'm ok with that too.

Reply:
I've made some changes, which I'll push for people to check out.
I went a bit beyond what's been suggested so far.